src/Microsoft.AspNetCore.Hosting/Internal/HostingApplicationDiagnostics.cs

```diff
@@ -46,12 +46,22 @@ public void BeginRequest(HttpContext httpContext, ref HostingApplication.Context
                 RecordRequestStartEventLog(httpContext);
             }
 
+            var diagnosticListenerEnabled = _diagnosticListener.IsEnabled();
+            var loggingEnabled = _logger.IsEnabled(LogLevel.Critical);
 
-            if (_diagnosticListener.IsEnabled())
+            // If logging is enabled or the diagnostic listener is enabled, try to get the correlation
+            // id from the header
+            StringValues correlationId;
+            if (diagnosticListenerEnabled || loggingEnabled)
+            {
+                httpContext.Request.Headers.TryGetValue(RequestIdHeaderName, out correlationId);
```

Comment:
We might want to add this as a known header to kestrel since we'll be doing it all the time (if logging is enabled).